designs/data_serialization_design.md

```diff
@@ -0,0 +1,189 @@
+
+
+## Requirements
+### Functional
+
+- **Complete**: Serialize the full data structure, including:
+    - Multi-level bucket data (probabilities, timestamps)
+    -  Configuration parameters
+    - MurmurHash seed
+    - Metadata (ID, flags)
+
+- **Portable**: Work across platforms, Go versions, and FAIR versions, with proper versioning.
+- **Compact**: Add minimal overhead for efficient transfer.
+- **Future-proof**: Support schema evolution as new parameters are added.
+
+### Non-Functional
+- **Performance**: Fast serialization/deserialization.
+- **Reliability**: Strong validation and error handling.
+- **Maintainability**: Clear versioning and schema upgrade path.
+- **Compatibility**: Backward and forward compatibility.
+
+## Serialization Format Selection
+
+Rejected options: JSON, XML, and other text formats — too verbose. \
+Custom serialization: Considered, but unnecessary given current requirements. Can revisit if protobuf overhead becomes a concern. \
+Chosen option: Protocol Buffers (protobuf), due to:
+- Binary format
+- Native support in gRPC
+- Strong typing and schema evolution
+- Built-in JSON support (for debugging)
+- Schema Design
+
+The schema has two broad parts:
+1)  Config (TrackerCfg) — defined by customers, drives tracker creation.
+2) Runtime Data — driven by traffic and execution:
+   - Consts: Initialized once, updated periodically (e.g., hash seed).
+   - State: Updated frequently on the hot path (bucket data).
+
+Accessing `State` requires the matching `Config` and `Consts`. Multiple tracker configs may exist simultaneously (multi-tenant).
+
+## Schema v1 (High-level)
+```
+package fair.data.v1;
+
+enum LevelSquashingFunction {
+  LEVEL_SQUASHING_FUNCTION_UNSPECIFIED = 0;
+  LEVEL_SQUASHING_FUNCTION_MIN = 1;
+  LEVEL_SQUASHING_FUNCTION_MEAN = 2;
+}
+
+enum HostSquashingFunction {
+  HOST_SQUASHING_FUNCTION_UNSPECIFIED = 0;
+  HOST_SQUASHING_FUNCTION_LOGIC1 = 1;
+  HOST_SQUASHING_FUNCTION_LOGIC2 = 2;
+}
+
+message TrackerCfg {
+  int64 config_guid = 1; // Unique per tracker config
+  int64 config_version = 2; // Monotonically increments per version of tracker config
+  uint32 m = 3;
+  uint32 l = 4;
+  double pi = 5;
+  double pd = 6;
+  double lambda = 7;
+  google.protobuf.Duration rotation_frequency = 8;
+  LevelSquashingFunction level_squash_fn = 9;
+  HostSquashingFunction host_squash_fn = 10;
+}
+
+message Bucket {
+  double probability = 1;
+  uint64 last_updated_time_ms = 2;
+}
+
+message Level {
+  repeated Bucket buckets = 1;
+}
+
+message FairData {
+  repeated Level levels = 1;
+}
+
+message FairRunConsts {
+  uint32 murmur_seed = 1;
+  string ref_guid = 2;
+}
+
+message HostMeta {
+  string host_guid = 1;
+  uint64 serialized_at_ms = 2;
+}
+
+message FairRunTimeData {
```

Comment:
What's this?

Reply:
with `FairRunTimeData` the  idea is to separate the config from the generated data (one time during rotation) and the collected client data (bucket data).  This way, we only choose to Serialize the Runtime data and not serialize the Configs.

What do you think?